### PR TITLE
Add ConfigSourceNIC indirection and implementation

### PR DESCRIPTION
**Pull request by manadart**
This patch broadly prepares for us to use different implementations for acquiring network configuration on-machine.

In particular:
- `ParseInterfaceType` and `GetBridgePorts` are implemented on `ConfigSource`. At the moment, these just call through to the methods that were previously accessed from `api/common`, but we are now at liberty to change them for specific implementations.
- Instead of `ConfigSource.InterfaceAddresses`, a new indirection for network interfaces, `ConfigSourceNIC` returns addresses. Relevant logic from `api/common` has been pushed into this type.
- Code relocated to `core/network` tests implementation detail as appropriate, with the tests in `api/common` now relying on mocks for the new indirections. This means new `ConfigSource` implementations no longer required changes in that package.

## QA steps

Despite the upheaval, there should be no functional changes.
- Bootstrap to LXD.
- `juju add-machine --series bionic`
- `juju add-machine lxd:0 --series bionic`
- Once quiesced, connect to Mongo and check that link-layer device discovery is working:
```
juju:PRIMARY> db.linklayerdevices.find().pretty()
{
        "_id" : "0ea16223-d982-4743-8520-bc35b68db2d6:m#0#d#lo",
        "name" : "lo",
        "model-uuid" : "0ea16223-d982-4743-8520-bc35b68db2d6",
        "mtu" : 65536,
        "machine-id" : "0",
        "type" : "loopback",
        "mac-address" : "",
        "is-auto-start" : true,
        "is-up" : true,
        "parent-name" : "",
        "virtual-port-type" : "",
        "txn-revno" : NumberLong(2),
        "txn-queue" : [
                "6038e03bb364523ca24fbb23_5f81380b"
        ]
}
{
        "_id" : "0ea16223-d982-4743-8520-bc35b68db2d6:m#0#d#eth0",
        "name" : "eth0",
        "model-uuid" : "0ea16223-d982-4743-8520-bc35b68db2d6",
        "mtu" : 1500,
        "machine-id" : "0",
        "type" : "ethernet",
        "mac-address" : "00:16:3e:93:c5:e0",
        "is-auto-start" : true,
        "is-up" : true,
        "parent-name" : "",
        "virtual-port-type" : "",
        "txn-revno" : NumberLong(3),
        "txn-queue" : [
                "6038e046b364523ca24fbcc4_1dfe9764"
        ],
        "providerid" : "nic-00:16:3e:93:c5:e0"
}
{
        "_id" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f:m#0#d#lo",
        "name" : "lo",
        "model-uuid" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f",
        "mtu" : 65536,
        "machine-id" : "0",
        "type" : "loopback",
        "mac-address" : "",
        "is-auto-start" : true,
        "is-up" : true,
        "parent-name" : "",
        "virtual-port-type" : "",
        "txn-revno" : NumberLong(2),
        "txn-queue" : [
                "6038e135b364523ca24fbcee_1dcf322b"
        ]
}
{
        "_id" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f:m#0#d#eth0",
        "name" : "eth0",
        "model-uuid" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f",
        "mtu" : 1500,
        "machine-id" : "0",
        "type" : "ethernet",
        "mac-address" : "00:16:3e:c3:a5:06",
        "is-auto-start" : true,
        "is-up" : true,
        "parent-name" : "",
        "virtual-port-type" : "",
        "txn-revno" : NumberLong(3),
        "txn-queue" : [
                "6038e13db364523ca24fbd45_82ee0cad"
        ],
        "providerid" : "nic-00:16:3e:c3:a5:06"
}
{
        "_id" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f:m#0#d#lxdbr0",
        "name" : "lxdbr0",
        "model-uuid" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f",
        "mtu" : 1500,
        "machine-id" : "0",
        "type" : "bridge",
        "mac-address" : "c2:32:52:bb:d8:0a",
        "is-auto-start" : true,
        "is-up" : true,
        "parent-name" : "",
        "virtual-port-type" : "",
        "txn-revno" : NumberLong(3),
        "txn-queue" : [
                "6038e3c4b364523ca24fbe8b_27c4c307"
        ],
        "providerid" : "nic-c2:32:52:bb:d8:0a"
}
{
        "_id" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f:m#0/lxd/0#d#lo",
        "name" : "lo",
        "model-uuid" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f",
        "mtu" : 65536,
        "machine-id" : "0/lxd/0",
        "type" : "loopback",
        "mac-address" : "",
        "is-auto-start" : true,
        "is-up" : true,
        "parent-name" : "",
        "virtual-port-type" : "",
        "txn-revno" : NumberLong(2),
        "txn-queue" : [
                "6038e3fab364523ca24fbec1_2956fbfb"
        ]
}
{
        "_id" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f:m#0/lxd/0#d#eth0",
        "name" : "eth0",
        "model-uuid" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f",
        "mtu" : 1500,
        "machine-id" : "0/lxd/0",
        "type" : "ethernet",
        "mac-address" : "00:16:3e:db:bf:40",
        "is-auto-start" : true,
        "is-up" : true,
        "parent-name" : "",
        "virtual-port-type" : "",
        "txn-revno" : NumberLong(2),
        "txn-queue" : [
                "6038e3fab364523ca24fbec1_2956fbfb"
        ]
}

juju:PRIMARY> db.ip.addresses.find().pretty()
{
        "_id" : "0ea16223-d982-4743-8520-bc35b68db2d6:m#0#d#lo#ip#127.0.0.1",
        "model-uuid" : "0ea16223-d982-4743-8520-bc35b68db2d6",
        "device-name" : "lo",
        "machine-id" : "0",
        "subnet-cidr" : "127.0.0.0/8",
        "config-method" : "loopback",
        "value" : "127.0.0.1",
        "origin" : "machine",
        "txn-revno" : NumberLong(2),
        "txn-queue" : [
                "6038e03bb364523ca24fbb23_5f81380b"
        ]
}
{
        "_id" : "0ea16223-d982-4743-8520-bc35b68db2d6:m#0#d#lo#ip#::1",
        "model-uuid" : "0ea16223-d982-4743-8520-bc35b68db2d6",
        "device-name" : "lo",
        "machine-id" : "0",
        "subnet-cidr" : "::1/128",
        "config-method" : "loopback",
        "value" : "::1",
        "origin" : "machine",
        "txn-revno" : NumberLong(2),
        "txn-queue" : [
                "6038e03bb364523ca24fbb23_5f81380b"
        ]
}
{
        "_id" : "0ea16223-d982-4743-8520-bc35b68db2d6:m#0#d#eth0#ip#10.62.163.252",
        "model-uuid" : "0ea16223-d982-4743-8520-bc35b68db2d6",
        "device-name" : "eth0",
        "machine-id" : "0",
        "subnet-cidr" : "10.62.163.0/24",
        "config-method" : "static",
        "value" : "10.62.163.252",
        "gateway-address" : "10.62.163.1",
        "is-default-gateway" : true,
        "origin" : "machine",
        "txn-revno" : NumberLong(3),
        "txn-queue" : [
                "6038e046b364523ca24fbcc4_1dfe9764"
        ],
        "provider-network-id" : "net-lxdbr0",
        "provider-subnet-id" : "subnet-lxdbr0-10.62.163.0/24"
}
{
        "_id" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f:m#0#d#lo#ip#127.0.0.1",
        "model-uuid" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f",
        "device-name" : "lo",
        "machine-id" : "0",
        "subnet-cidr" : "127.0.0.0/8",
        "config-method" : "loopback",
        "value" : "127.0.0.1",
        "origin" : "machine",
        "txn-revno" : NumberLong(2),
        "txn-queue" : [
                "6038e135b364523ca24fbcee_1dcf322b"
        ]
}
{
        "_id" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f:m#0#d#lo#ip#::1",
        "model-uuid" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f",
        "device-name" : "lo",
        "machine-id" : "0",
        "subnet-cidr" : "::1/128",
        "config-method" : "loopback",
        "value" : "::1",
        "origin" : "machine",
        "txn-revno" : NumberLong(2),
        "txn-queue" : [
                "6038e135b364523ca24fbcee_1dcf322b"
        ]
}
{
        "_id" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f:m#0#d#eth0#ip#10.62.163.47",
        "model-uuid" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f",
        "device-name" : "eth0",
        "machine-id" : "0",
        "subnet-cidr" : "10.62.163.0/24",
        "config-method" : "static",
        "value" : "10.62.163.47",
        "gateway-address" : "10.62.163.1",
        "is-default-gateway" : true,
        "origin" : "machine",
        "txn-revno" : NumberLong(3),
        "txn-queue" : [
                "6038e13db364523ca24fbd45_82ee0cad"
        ],
        "provider-network-id" : "net-lxdbr0",
        "provider-subnet-id" : "subnet-lxdbr0-10.62.163.0/24"
}
{
        "_id" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f:m#0#d#lxdbr0#ip#10.57.118.1",
        "model-uuid" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f",
        "device-name" : "lxdbr0",
        "machine-id" : "0",
        "subnet-cidr" : "10.57.118.0/24",
        "config-method" : "static",
        "value" : "10.57.118.1",
        "origin" : "machine",
        "txn-revno" : NumberLong(3),
        "txn-queue" : [
                "6038e3c4b364523ca24fbe8b_27c4c307"
        ],
        "provider-subnet-id" : "subnet--10.57.118.0/24"
}
{
        "_id" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f:m#0/lxd/0#d#lo#ip#127.0.0.1",
        "model-uuid" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f",
        "device-name" : "lo",
        "machine-id" : "0/lxd/0",
        "subnet-cidr" : "127.0.0.0/8",
        "config-method" : "loopback",
        "value" : "127.0.0.1",
        "origin" : "machine",
        "txn-revno" : NumberLong(2),
        "txn-queue" : [
                "6038e3fab364523ca24fbec1_2956fbfb"
        ]
}
{
        "_id" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f:m#0/lxd/0#d#lo#ip#::1",
        "model-uuid" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f",
        "device-name" : "lo",
        "machine-id" : "0/lxd/0",
        "subnet-cidr" : "::1/128",
        "config-method" : "loopback",
        "value" : "::1",
        "origin" : "machine",
        "txn-revno" : NumberLong(2),
        "txn-queue" : [
                "6038e3fab364523ca24fbec1_2956fbfb"
        ]
}
{
        "_id" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f:m#0/lxd/0#d#eth0#ip#10.57.118.226",
        "model-uuid" : "8ad8de94-f3ff-4ee0-86f4-ead38923f83f",
        "device-name" : "eth0",
        "machine-id" : "0/lxd/0",
        "subnet-cidr" : "10.57.118.0/24",
        "config-method" : "static",
        "value" : "10.57.118.226",
        "gateway-address" : "10.57.118.1",
        "is-default-gateway" : true,
        "origin" : "machine",
        "txn-revno" : NumberLong(2),
        "txn-queue" : [
                "6038e3fab364523ca24fbec1_2956fbfb"
        ]
}
```

## Documentation changes

None.

## Bug reference

Advances https://bugs.launchpad.net/juju/+bug/1863916

### CODE REVIEW

File: core/network/source_other.go [hunk 1]
```diff
@@ -38,9 +112,13 @@ func NewNetAddr(a string) *netAddr {
 	if ip == nil {
 		ip = net.ParseIP(a)
 	}
-	res.ip = ip
 
-	return res
+	if ip == nil {
```
Comment (manadart) [inline core/network/source_other.go:116]: This checking for validity was requested by @achilleasa in a pre-cursor to this patch.

File: api/common/network.go [hunk 2]
```diff
@@ -62,26 +59,26 @@ func GetObservedNetworkConfig(source corenetwork.ConfigSource) ([]params.Network
...
 	for _, nic := range interfaces {
-		virtualPortType := corenetwork.NonVirtualPort
-		if knownOVSBridges.Contains(nic.Name) {
-			virtualPortType = corenetwork.OvsPort
+		virtualPortType := network.NonVirtualPort
+		if knownOVSBridges.Contains(nic.Name()) {
+			virtualPortType = network.OvsPort
 		}
 
-		nicType := network.ParseInterfaceType(sysClassNetPath, nic.Name)
```
Comment (manadart) [inline api/common/network.go:72]: Note this now pushed into the implementation.

File: api/common/network.go [hunk 3]
```diff
@@ -162,9 +153,10 @@ func interfaceToNetworkConfig(nic net.Interface,
 	}
 }
 
-func updateParentForBridgePorts(bridgeName, sysClassNetPath string, nameToConfigs map[string][]params.NetworkConfig) {
-	ports := network.GetBridgePorts(sysClassNetPath, bridgeName)
```
Comment (manadart) [inline api/common/network.go:166]: And this.

File: api/common/network.go [hunk 4]
```diff
@@ -62,26 +59,26 @@ func GetObservedNetworkConfig(source corenetwork.ConfigSource) ([]params.Network
 	}
 	var namesOrder []string
 	nameToConfigs := make(map[string][]params.NetworkConfig)
-	sysClassNetPath := source.SysClassNetPath()
```
Comment (manadart) [inline api/common/network.go:65]: Which means this need not be exposed.
The leaks in the abstraction are now plugged.

File: core/network/source.go [hunk 5]
```diff
@@ -58,4 +80,80 @@ type ConfigSource interface {
...
+	const deviceType = "DEVTYPE="
+	location := filepath.Join(sysPath, interfaceName, "uevent")
+
+	data, err := ioutil.ReadFile(location)
+	if err != nil {
+		logger.Debugf("ignoring error reading %q: %v", location, err)
+		return UnknownInterface
+	}
+
+	devtype := ""
```
Comment (SimonRichardson) [inline core/network/source.go:106]:
```suggestion
	var devtype string
```

File: core/network/source.go [hunk 6]
```diff
@@ -58,4 +80,80 @@ type ConfigSource interface {
...
+// InterfaceType. SysClassNetPath should be passed as sysPath. Returns
+// UnknownInterface if the type cannot be reliably determined for any reason.
+// Example call: network.ParseInterfaceType(network.SysClassNetPath, "br-eth1")
+// TODO (manadart 2021-02-12): As with SysClassNetPath above, specific
+// implementations should be sought for this that are OS-dependent.
+func ParseInterfaceType(sysPath, interfaceName string) InterfaceType {
+	const deviceType = "DEVTYPE="
+	location := filepath.Join(sysPath, interfaceName, "uevent")
+
+	data, err := ioutil.ReadFile(location)
```
Comment (SimonRichardson) [inline core/network/source.go:100]: Is it wise to read the whole file? Instead, we could use a bufio.NewScanner and read line by line.

File: core/network/source.go [hunk 7]
```diff
@@ -58,4 +80,80 @@ type ConfigSource interface {
...
+// Example call: network.GetBridgePorts(network.SysClassNetPath, "br-eth1")
+// TODO (manadart 2021-02-12): As with SysClassNetPath above, specific
+// implementations should be sought for this that are OS-dependent.
+func GetBridgePorts(sysPath, bridgeName string) []string {
+	portsGlobPath := filepath.Join(sysPath, bridgeName, "brif", "*")
+	// Glob ignores I/O errors and can only return ErrBadPattern, which we treat
+	// as no results, but for debugging we're still logging the error.
+	paths, err := filepath.Glob(portsGlobPath)
+	if err != nil {
+		logger.Debugf("ignoring error traversing path %q: %v", portsGlobPath, err)
```
Comment (SimonRichardson) [inline core/network/source.go:145]: Why not just `return nil` here, as paths will be empty anyway?

File: core/network/source_other.go [hunk 8]
```diff
@@ -59,43 +137,28 @@ func (a *netAddr) String() string {
...
-	result := make([]ConfigSourceAddr, 0, len(addrs))
-	for _, addr := range addrs {
-		if addr.String() != "" {
-			result = append(result, NewNetAddr(addr.String()))
-		}
+	result := make([]ConfigSourceNIC, len(nics))
+	for i := range nics {
+		// Close over the sysClassNetPath so that
+		// the NIC needs to know nothing about it.
+		parseType := func(name string) InterfaceType { return ParseInterfaceType(n.sysClassNetPath, name) }
```
Comment (SimonRichardson) [inline core/network/source_other.go:155]: Lift this outside of the for loop?

Comment (manadart) [reply core/network/source_other.go:155]: Yes. Good catch.

File: core/network/source.go (hunk 7 above)
Comment (manadart) [reply core/network/source.go:145]: The next line achieves this - we always return nil for an empty path.

File: core/network/source.go (hunk 6 above)
Comment (manadart) [reply core/network/source.go:100]: The file is 2 or 3 lines, so it's no big deal.

File: core/network/source.go (hunk 5 above)
Comment (manadart) [reply core/network/source.go:106]: Done.

**Review by SimonRichardson — APPROVED** [review]

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**